File: pylors_asset_mgr/core/workspace.py
from pathlib import Path
from .tag_store import TagStore


class Workspace:
    """管理一个工作区 — 真实文件系统目录 + 其 .pylorsmeta.db。"""

    def __init__(self, root_path: Path | str):
        self.root = Path(root_path).resolve()
        self.db_path = self.root / ".pylorsmeta.db"
        self.tag_store = TagStore(self.db_path)
        self.display_name = self.root.name

    def initialize(self):
        """确保 DB 文件存在并初始化 schema。"""
        self.root.mkdir(parents=True, exist_ok=True)
        self.tag_store.initialize_database()

    def close(self):
        self.tag_store.close()

    def is_valid(self) -> bool:
        return self.root.exists() and self.root.is_dir()

    # --- 标签操作委托 ---

    def get_user_tags(self, rel_path: str) -> dict[str, str]:
        return self.tag_store.get_all_tags(rel_path)

    def set_user_tag(self, rel_path: str, key: str, value: str):
        self.tag_store.set_tag(rel_path, key, value)

    def delete_user_tag(self, rel_path: str, key: str):
        self.tag_store.delete_tag(rel_path, key)

    def get_all_tag_keys(self) -> list[str]:
        return self.tag_store.get_all_keys()

    def batch_set_tags(self, rel_paths: list[str], key: str, value: str):
        self.tag_store.batch_set_tags(rel_paths, key, value)

    def get_config(self, key: str, default: str = "") -> str:
        return self.tag_store.get_config(key, default)

    def set_config(self, key: str, value: str):
        self.tag_store.set_config(key, value)

    def cleanup_stale_tags(self, valid_paths: set[str]):
        self.tag_store.delete_tags_for_missing_files(valid_paths)

File: PylorsAssetMgr/Models/Workspace.swift
import Foundation

actor Workspace {
    let root: URL
    let dbPath: URL
    let tagStore: TagStore
    let displayName: String

    init(rootPath: String) throws {
        self.root = URL(fileURLWithPath: rootPath).standardizedFileURL
        self.dbPath = root.appendingPathComponent(".pylorsmeta.db")
        self.tagStore = try TagStore(dbPath: dbPath.path)
        self.displayName = root.lastPathComponent
    }

    var isValid: Bool {
        let fm = FileManager.default
        var isDir: ObjCBool = false
        return fm.fileExists(atPath: root.path, isDirectory: &isDir) && isDir.boolValue
    }

    // MARK: - 标签操作

    func getUserTags(relPath: String) async throws -> [String: String] {
        try await tagStore.getAllTags(filePath: relPath)
    }

    func setUserTag(relPath: String, key: String, value: String) async throws {
        try await tagStore.setTag(filePath: relPath, key: key, value: value)
    }

    func deleteUserTag(relPath: String, key: String) async throws {
        try await tagStore.deleteTag(filePath: relPath, key: key)
    }

    func getAllTagKeys() async throws -> [String] {
        try await tagStore.getAllKeys()
    }

    func batchSetTags(relPaths: [String], key: String, value: String) async throws {
        try await tagStore.batchSetTags(filePaths: relPaths, key: key, value: value)
    }

    func cleanupStaleTags(validPaths: Set<String>) async throws {
        try await tagStore.cleanupStaleTags(validPaths: validPaths)
    }

    // MARK: - 配置

    func getConfig(_ key: String, default: String = "") async throws -> String {
        try await tagStore.getConfig(key, default: `default`)
    }

    func setConfig(_ key: String, value: String) async throws {
        try await tagStore.setConfig(key, value: value)
    }

    // MARK: - 键管理

    func registerKey(_ key: String) async throws {
        try await tagStore.registerKey(key)
    }

    func deleteKey(_ key: String) async throws {
        try await tagStore.deleteKey(key)
    }

    func renameKey(old: String, new: String) async throws {
        try await tagStore.renameKey(old: old, new: new)
    }

    func getDistinctValues(for key: String) async throws -> [String] {
        try await tagStore.getDistinctValues(for: key)
    }

    func deleteValue(key: String, value: String) async throws {
        try await tagStore.deleteValue(key: key, value: value)
    }
}
